res/css/views/context_menus/_IconizedContextMenu.scss

```diff
@@ -50,21 +50,21 @@ limitations under the License.
         }
 
         // round the top corners of the top button for the hover effect to be bounded
-        &:first-child .mx_AccessibleButton:first-child {
+        &:first-child .mx_AccessibleButton:not(.mx_AccessibleButton_hasKind):first-child {
```

Comment:
Avoid clashes when trying to put a primary accessible button inside of a `ContextMenu`

Reply:
There is a reason not to do that, it won't be accessible for keyboard-only users. Context Menus have strict focus management requirements

Reply:
(new better flow suggestions welcome) 🙂

Popping off to a modal feels a little heavy.

Tabbing to a date picker in the context menu seems in the realm of sane to handle for the keyboard. But the HTML5 date picker here isn't giving the best UX because ideally we could remove the "Go" button and "Go" automatically when a date is picked. This doesn't seem feasible though because there no way to distinguish a `change` from navigating the months of the picker vs a final selection.

Reply:
Tab in a context menu closes the context menu though as it is focus-locked

Reply:
I think we can make this work. And seems acceptable under ARIA guidelines:

>  - `Tab`: Moves focus to the next element in the tab sequence, and if the item that had focus is not in a `menubar`, closes its `menu` and all open parent `menu` containers.
>
>  - `Shift + Tab`: Moves focus to the previous element in the tab sequence, and if the item that had focus is not in a `menubar`, closes its `menu` and all open parent `menu` containers.
>
> https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-12

---

I had this mostly working in https://github.com/matrix-org/matrix-react-sdk/pull/7317/commits/9fa980c16d0ae396ac830ec72d0e6a6290477d1c except for closing the menu, but tab was navigating the items as expected.

I think I will try to adapt this to use the `RovingTabIndex` like a `TODO` comment in `ContextMenu` mentions so I can register the items explicitly for navigation and be able to tell when we tab out the end or the start to close the menu.

Reply:
Is there a `RovingTabIndex` tab index that is known working and well implemented that I can base this off of? The spaces panel seems like a pretty similar with the tree view being navigable with arrow keys and can `Tab` to the space options of each as you go.

The `MessageActionBar` which uses `Toolbar` and `RovingTabIndex` seems broken as I can't figure out how to navigate by `Tab` or `Arrow` keys in it.

![](https://user-images.githubusercontent.com/558581/145537495-dde32259-d950-4e89-8600-4ad2bda54f21.png)

It also seems counter-intuitive to me that I'm unable to navigate menus by `Tab` and have to use `Arrow` keys but it also seems inconsistent 🤔. Both seem like they should work.

What's the goal of setting `tabindex="-1"` on elements which are focusable?

Reply:
> It also seems counter-intuitive to me that I'm unable to navigate menus by Tab and have to use Arrow keys but it also seems inconsistent 🤔. Both seem like they should work.

> What's the goal of setting tabindex="-1" on elements which are focusable? 

See https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets Technique 1

It is so the app doesn't have a million tabstops, and instead each widget is a tabstop with sane and expected keyboard & focus management, e.g arrows. The roving & treeview stuff was written to match the wai-aria practices

Reply:
`MessageActionBar` does seem to have regressed.
The space panel, room list, space home are all roving focus managed

Reply:
https://github.com/matrix-org/matrix-react-sdk/pull/7336 fixes MAB, during a refactoring something got missed I think

Reply:
Thanks for looking into the fixes for the `MessageActionBar `! I've created a v2 PR, https://github.com/matrix-org/matrix-react-sdk/pull/7339, based off of those fixes and wanting to refactor `ContextMenu` to use `RovingTabIndex` so that I can leave this PR untouched in its somewhat working state.